### PR TITLE
Add an error field to Record

### PR DESCRIPTION
**Pull request by KodrAus**
For #357 

Adds a `dyn Error + 'static` field to `Record` so that concrete errors can be associated with a log event. I'd thought about adding a `Send` requirement, but decided against it because `Record` already isn't `Send`, there are error types out there folks might want to log that aren't `Send`, and since you can downcast errors if your app needs to send log records to other threads you could use that to downcast and build your own sendable record type.

r? @sfackler 

# Future extension

We could extend the macro to allow something like this in caller code:

```rust
let o = do_some_work().map_err(|e| {
    error!(source: &e, target: "some target", "I tried so hard, and got so far");
    e
})?;
```

Macro support could be added in a fairly naive way to the log macros, or we could take another stab at a tt muncher to support arbitrary special fields before the arguments.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -816,6 +818,13 @@ impl<'a> Record<'a> {
         self.line
     }
 
+    /// The error associated with the message.
```
Comment (sfackler) [inline src/lib.rs]: This should mention it requires the std feature.

File: src/lib.rs [hunk 2]
```diff
@@ -979,6 +992,14 @@ impl<'a> RecordBuilder<'a> {
         self
     }
 
+    /// Set [`error`](struct.Record.html#method.error)
```
Comment (sfackler) [inline src/lib.rs]: Same here

**Comment by sfackler** [comment]
Seems reasonable to me to not have a Send bound - even if the fmt::Arguments field wasn't also there you can't really ship a Record around since the error is just borrowed.

I do think we should figure out what the macro change looks like before cutting a release with this, though.

**Comment by KodrAus** [comment]
> I do think we should figure out what the macro change looks like before cutting a release with this, though.

Sure :+1: I might leave this PR sitting open then and switch over to #357 where we can work out how we want the macro to look. I don't think it'll be as risky to change the macros following the same scheme as `target` as the kv support was, because we're not messing with the token soup that gets forwarded to `format_args`.

**Comment by KodrAus** [comment]
This has fallen pretty out of date now so I'll close it and revisit in the future.

**Pull request closed without merging**